### ISSUE TITLE
[Bug]: ObservableGauge no longer dropping data points from old Attributes values

### ISSUE THREAD

**Issue by markdingram** [open-telemetry/opentelemetry-rust#2213]
### Describe the solution you'd like:

A change in behaviour is present between 0.24 & 0.26 for ObservableGauge & possibly other Async Instruments.

In 0.26 once any reading is made for a given combination of attributes that reading is always sent in subsequent collections. There's no way to remove any attributes that are no longer present in the data. This doesn't match the [SDK Spec](https://github.com/open-telemetry/opentelemetry-specification/blob/0a78571045ca1dca48621c9648ec3c832c3c541c/specification/metrics/sdk.md#metricreader):

> For asynchronous instruments with Delta or Cumulative aggregation temporality, [MetricReader.Collect](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#collect) MUST only receive data points with measurements recorded since the previous collection.

Adapted one of the existing tests to show an error - <https://github.com/markdingram/opentelemetry-rust/commit/68d375b320c1e2cbcec5030a2f0269cd9553a7fb>

Comment from @cijothomas on CNCF Slack: 

> OTel Rust currently is likely violating the above part of spec, as it never forgets data points for Cumulative temporality, but the spec says it should forget data points for CUMULATIVE too, for Observable instruments.

**Comment by kensimon** [comment]
I believe this bug was introduced here: https://github.com/open-telemetry/opentelemetry-rust/pull/2021

The LastValue aggregation function used to have a comment saying "[Builder::temporality] is ignored and delta is always used", and "Delta temporality is the only temporality that makes semantic sense for a last-value aggregate", but now it obeys self.temporality and will apply cumulative data if self.temporality is not Delta.

Interestingly the doc comments for AggregateBuilder still say that Delta temporality will be used for last_value: https://github.com/open-telemetry/opentelemetry-rust/blob/52f6d9214eca58c32e3fc41e3829038b868c07e1/opentelemetry-sdk/src/metrics/internal/aggregate.rs#L60

It would seem to me that the original comments were right... last_value should mean last value even if the AggregateBuilder's temporality is cumulative.

This is effecting me because the opentelemetry_prometheus's PrometheusExporter only supports Temporality::Cumulative: https://github.com/open-telemetry/opentelemetry-rust/blob/52f6d9214eca58c32e3fc41e3829038b868c07e1/opentelemetry-prometheus/src/lib.rs#L155

```
    /// Note: Prometheus only supports cumulative temporality so this will always be
    /// [Temporality::Cumulative].
```

Meaning that trying to use the prometheus exporter, there's no way to clear any gauge fields in v0.26.

**Comment by cijothomas** [comment]
Thanks for reporting! Yes, this looks like a bug.

(Co-incidentally, this bug was also discovered in OTel .NET recently : https://github.com/open-telemetry/opentelemetry-dotnet/pull/5952)

**Comment by etep** [comment]
Is it possible to provide an API that clears out attributes that have not been touched in some amount of time? We hit this bug too... we noticed it because of an increase in data rate to our observability cloud service provider. Basically, we were leaking attributes. Our fix was to stop using an attribute, but this is sub-optimal.

Also - not clear from this comment thread, but what is the scope of attribute leakage? Before finding this bug report, we thought the scope as all of rust+otel, not just ObservableGauge.

**Comment by cijothomas** [comment]
@etep Not sure I follow the issue you are reported. This issue is accepted as a bug, and we'll be fixing it.

Regarding clearing out attributes that are unused for some point of time - if you are using Delta, this is already the case. Every collection (exporter) will reset the previous state, so attributes that are not used in that period will get cleared out.
For cumulative, we do not ever forget. And this is by-design for cumulative. It is technically possible for us to offer a solution to start forgetting even in Cumulative and reset the start time, but such capability is not planned. 

Could you elaborate more on the issue you are facing?

**Comment by etep** [comment]
Hi @cijothomas - we did not realize that the default temporality selector was cumulative and that it would leak attributes. Leaking attributes is definitely not anticipated behavior.

I think that `DefaultTemporalitySelector` is not available if we update our rust otel version. Given that we did not make any kind of conscientious choice here, we'll need to think about it.

Rolling back to when I wrote the comment: we are just responding to unanticipated behavior.

**Comment by cijothomas** [comment]
> I think that DefaultTemporalitySelector is not available if we update our rust otel version

Temporality is configured in new Otel as well. Its done differently, but entirely possible.
https://github.com/open-telemetry/opentelemetry-rust/blob/main/examples/metrics-basic/src/main.rs#L9-L10

**Comment by markdingram** [comment]
> > I think that DefaultTemporalitySelector is not available if we update our rust otel version
> 
> Temporality is configured in new Otel as well. Its done differently, but entirely possible. https://github.com/open-telemetry/opentelemetry-rust/blob/main/examples/metrics-basic/src/main.rs#L9-L10


Switching Temporality is a workaround, but does come with tradeoffs outlined - https://grafana.com/blog/2023/09/26/opentelemetry-metrics-a-guide-to-delta-vs.-cumulative-temporality-trade-offs/

**Comment by markdingram** [comment]
There's a nice example in the OTEL docs showing the expected behaviour for as Async Instrument - https://opentelemetry.io/docs/specs/otel/metrics/supplementary-guidelines/#asynchronous-example-cumulative-temporality - check the outputs for `PID1001`

Could one of the maintainers could write some guidance on where / how to tackle this? If its a small change I could help and/or contribute some tests.

**Comment by cijothomas** [comment]
> Could one of the maintainers could write some guidance on where / how to tackle this? If it's a small change I could help and/or contribute some tests.

Yes. Metric aggregation uses a HashMap to keep track of all KVP combinations.

When using Cumulative, the HashMap is read by the Collect() thread and exports everything it finds in the HashMap. As HashMap is not drained, every KVP once in the HashMap will be reported forever.

When using Delta, the HashMap is *drained* and exports everything it finds in the HashMap. i.e After each Collect(), the hashMap is reset, so unless a KVP is reported again in new cycle, it won't be exported.

The bug is in the Cumulative.
When using ObservableCounter/Gauges, spec says, only those KVP which has reported in a given cycle should be exported. In other words, we need to introduce some "clearing" up in Cumulative, when the instrument used in Observables. 

https://github.com/open-telemetry/opentelemetry-rust/blob/main/opentelemetry-sdk/src/metrics/internal/mod.rs is where most of the logic resides. @fraillt has done a lot of optimizations in this area (still doing! pending review), and might be the best person to help as well.

**Comment by markdingram** [comment]
Thanks! I see ObservableGauge should now have prior behaviour, will give this a try.
